server/e2e_test.go

```diff
@@ -331,35 +331,32 @@ func newTestInbox(conn net.Conn) *testInbox {
 // PopOldest pops the oldest message and returns it or returns an error
 // if no message appears within a short time period.
 func (inbox *testInbox) PopOldest() (*packetInfo, error) {
+	var packet *packetInfo
+
 	err := retry.Do(
 		func() error {
 			inbox.mutex.Lock()
+			defer inbox.mutex.Unlock()
 			if len(inbox.packets) == 0 {
-				inbox.mutex.Unlock()
 				return fmt.Errorf("empty inbox")
 			}
-			// retain the lock if we find a message
+			packet = inbox.packets[0]
+			inbox.packets = inbox.packets[1:]
 			return nil
 		},
 		retry.Attempts(100),
-		retry.Delay(1*time.Millisecond),
+		retry.Delay(5*time.Millisecond),
 	)
-	if err != nil {
-		panic(err)
-	}
-	defer inbox.mutex.Unlock()
-
-	p := inbox.packets[0]
-	inbox.packets = inbox.packets[1:]
-	return p, nil
+	return packet, err
 }
 
-// AssertRegistered confirms registration, consumes expected registration
-// message, and returns registration details
-func (h *testHarness) AssertRegistered(c *testClient) (uint32, []byte) {
+// WaitRegistered waits for and consumes the registration response
+// and returns registration details.
+func (h *testHarness) WaitRegistered(c *testClient) (uint32, []byte) {
+	// popping the message is where the wait happens
```

Comment:
Lots of renaming to clarify which methods do waiting and which ones act without waiting for anything.